scipy/stats/_axis_nan_policy.py

```diff
@@ -0,0 +1,347 @@
+    xs = [np.broadcast_to(x, shape + (x.shape[-1],)) for x in xs]
+    # concatenate along last axis
+    res = np.concatenate(xs, axis=-1)
+    # move the last axis back to where it was
+    res = np.swapaxes(res, axis, -1)
+    return res
+
+
+# TODO: add support for `axis` tuples
+def _remove_nans(samples, paired):
+    "Remove nans from paired or unpaired samples"
+    # potential optimization: don't copy arrays that don't contain nans
+    if not paired:
+        return [sample[~np.isnan(sample)] for sample in samples]
+
+    # for paired samples, we need to remove the whole pair when any part
+    # has a nan
+    nans = np.isnan(samples[0])
+    for sample in samples[1:]:
+        nans = nans | np.isnan(sample)
+    not_nans = ~nans
+    return [sample[not_nans] for sample in samples]
+
+
+def _check_empty_inputs(samples, axis):
+    """
+    Check for empty sample; return appropriate output for a vectorized hypotest
+    """
+    # if none of the samples are empty, we need to perform the test
+    if not any((sample.size == 0 for sample in samples)):
+        return None
+    # otherwise, the statistic and p-value will be either empty arrays or
+    # arrays with NaNs. Produce the appropriate array and return it.
+    output_shape = _broadcast_array_shapes_remove_axis(samples, axis)
+    output = np.ones(output_shape) * np.nan
+    return output
+
+
+# Standard docstring / signature entries for `axis` and `nan_policy`
+_name = 'axis'
+_type = "int or None, default: 0"
+_desc = (
+    """If an int, the axis of the input along which to compute the statistic.
+    The statistic of each axis-slice (e.g. row) of the input will appear in a
+    corresponding element of the output.
+    If ``None``, the input will be raveled before computing the statistic."""
+    .split('\n'))
+_axis_parameter_doc = Parameter(_name, _type, _desc)
+_axis_parameter = inspect.Parameter(_name,
+                                    inspect.Parameter.KEYWORD_ONLY,
+                                    default=0)
+
+_name = 'nan_policy'
+_type = "{'propagate', 'omit', 'raise'}"
+_desc = (
+    """Defines how to handle input NaNs.
+
+    - ``propagate``: if a NaN is present in the axis slice (e.g. row) along
+      which the  statistic is computed, the corresponding entry of the output
+      will be NaN.
+    - ``omit``: NaNs will be omitted when performing the calculation.
+      If insufficient data remains in the axis slice along which the
+      statistic is computed, the corresponding entry of the output will be
+      NaN.
+    - ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
+    .split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _axis_nan_policy_factory(result_object, default_axis=0,
```

Comment:
Appearances of `_vectorize_hypotest` have been changed to `_axis_nan_policy` as this also adds `nan_policy` and it is not specific to hypothesis tests. Other related refactorings have been done in the test suite.